[RFC] Profile rpc_async call from JIT

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40652 [RFC] Profile rpc_async call from JIT**

Resolves https://github.com/pytorch/pytorch/issues/40304, but looking for
feedback on whether there is a better approach for this (cc @ilia-cher).

In order to profile `rpc_async` calls made within a torchscript function, we
add the profiling logic to `rpcTorchscript` which is the point where the RPC is
dispatched and is called by the jit `rpc_async` operator. We take a somewhat
similar approach to how this is done in the python API. If profiling is
enabled, we call `record_function_enter` which creates a `RecordFunction`
object and runs its starting callbacks. Then, we schedule end callbacks for
this `RecordFunction` to be run when the jit future completes.

One caveat is that `rpcTorchscript` can also be called by rpc_async from a
non-JIT function, in which case the profiling logic lives in Python. We add a
check to ensure that we don't double profile in this case.

Differential Revision: [D22270608](https://our.internmc.facebook.com/intern/diff/D22270608/)

rohan-varma committed:
[RFC] Profile rpc_async call from JIT

Resolves https://github.com/pytorch/pytorch/issues/40304, but looking for
feedback on whether there is a better approach for this.

In order to profile `rpc_async` calls made within a torchscript function, we
add the profiling logic to `rpcTorchscript` which is the point where the RPC is
dispatched and is called by the jit `rpc_async` operator. We take a somewhat
similar approach to how this is done in the python API. If profiling is
enabled, we call `record_function_enter` which creates a `RecordFunction`
object and runs its starting callbacks. Then, we schedule end callbacks for
this `RecordFunction` to be run when the jit future completes.

One caveat is that `rpcTorchscript` can also be called by rpc_async from a
non-JIT function, in which case the profiling logic lives in Python. We add a
check to ensure that we don't double profile in this case.

Differential Revision: [D22270608](https://our.internmc.facebook.com/intern/diff/D22270608/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RFC] Profile rpc_async call from JIT"


Resolves https://github.com/pytorch/pytorch/issues/40304, but looking for
feedback on whether there is a better approach for this (cc @ilia-cher).

In order to profile `rpc_async` calls made within a torchscript function, we
add the profiling logic to `rpcTorchscript` which is the point where the RPC is
dispatched and is called by the jit `rpc_async` operator. We take a somewhat
similar approach to how this is done in the python API. If profiling is
enabled, we call `record_function_enter` which creates a `RecordFunction`
object and runs its starting callbacks. Then, we schedule end callbacks for
this `RecordFunction` to be run when the jit future completes.

One caveat is that `rpcTorchscript` can also be called by rpc_async from a
non-JIT function, in which case the profiling logic lives in Python. We add a
check to ensure that we don't double profile in this case.

Differential Revision: [D22270608](https://our.internmc.facebook.com/intern/diff/D22270608/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/torchscript_functions.cpp:
```diff
@@ -20,13 +21,21 @@ c10::intrusive_ptr<c10::ivalue::Future> rpcTorchscript(
+  at::Tensor handle = at::zeros(1);
+  auto shouldProfile = torch::autograd::profiler::profilerEnabled() &&
       !torch::distributed::rpc::RemoteProfilerManager::getInstance()
-           .isCurrentKeySet()) {
+           .isCurrentKeySet();
```
Is there a unit test that exercises the `isCurrentKeySet` condition?

rohan-varma replied:
Yes, since this code path is triggered both by `rpc_async` being called with a `ScriptFunction` in normal python mode, and also `rpc_async` call from a TorchScript function. In the former case, `isCurrentKeySet` is true since we set that in python (`rpc/api.py`). So any of the tests such as `test_profiler_with_script_async_rpc` will test the case where it's true, and the case where it's false is tested in the test added to this diff.

pritamdamania87 reviewed (commented):
LGTM, will let @ilia-cher stamp this though.

ilia-cher commented:
@pritamdamania this is mainly the RPC change as I understand, as the main change is in `rpcTorchscript` function which you guys should be more familiar with

rohan-varma commented:
@ilia-cher This is indeed mostly RPC changes, but I wanted to confirm with you regarding the use of `record_function_ops.record_function_enter` or whether there is a better way that you suggest instead.

rohan-varma committed:
Update on "[RFC] Profile rpc_async call from JIT"


Resolves https://github.com/pytorch/pytorch/issues/40304, but looking for
feedback on whether there is a better approach for this (cc @ilia-cher).

In order to profile `rpc_async` calls made within a torchscript function, we
add the profiling logic to `rpcTorchscript` which is the point where the RPC is
dispatched and is called by the jit `rpc_async` operator. We take a somewhat
similar approach to how this is done in the python API. If profiling is
enabled, we call `record_function_enter` which creates a `RecordFunction`
object and runs its starting callbacks. Then, we schedule end callbacks for
this `RecordFunction` to be run when the jit future completes.

One caveat is that `rpcTorchscript` can also be called by rpc_async from a
non-JIT function, in which case the profiling logic lives in Python. We add a
check to ensure that we don't double profile in this case.

Differential Revision: [D22270608](https://our.internmc.facebook.com/intern/diff/D22270608/)

[ghstack-poisoned]

ilia-cher commented:
I guess usage of _call_end_callbacks_on_fut is justified here as mirroring python's record_function, we could've added a C++ version of the guard too (up to you)

rohan-varma committed:
Update on "[RFC] Profile rpc_async call from JIT"


Resolves https://github.com/pytorch/pytorch/issues/40304, but looking for
feedback on whether there is a better approach for this (cc @ilia-cher).

In order to profile `rpc_async` calls made within a torchscript function, we
add the profiling logic to `rpcTorchscript` which is the point where the RPC is
dispatched and is called by the jit `rpc_async` operator. We take a somewhat
similar approach to how this is done in the python API. If profiling is
enabled, we call `record_function_enter` which creates a `RecordFunction`
object and runs its starting callbacks. Then, we schedule end callbacks for
this `RecordFunction` to be run when the jit future completes.

One caveat is that `rpcTorchscript` can also be called by rpc_async from a
non-JIT function, in which case the profiling logic lives in Python. We add a
check to ensure that we don't double profile in this case.

Differential Revision: [D22270608](https://our.internmc.facebook.com/intern/diff/D22270608/)

[ghstack-poisoned]
